test/extensions/transport_sockets/tls/handshaker_test.cc

```diff
@@ -0,0 +1,173 @@
+#include "envoy/network/transport_socket.h"
+#include "envoy/ssl/handshaker.h"
+#include "envoy/ssl/socket_state.h"
+
+#include "extensions/transport_sockets/tls/handshaker_impl.h"
+
+#include "test/extensions/transport_sockets/tls/ssl_certs_test.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "openssl/evp.h"
+#include "openssl/hmac.h"
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+namespace {
+
+using ::testing::StrictMock;
+
+// A callback shaped like pem_password_cb.
+// See https://www.openssl.org/docs/man1.1.0/man3/pem_password_cb.html.
+int PemPasswordCallback(char* buf, int buf_size, int /*rwflag*/, void* u) {
+  if (u == nullptr) {
+    return 0;
+  }
+  std::string passphrase = *reinterpret_cast<std::string*>(u);
+  ASSERT(buf_size >= static_cast<int>(passphrase.size()));
+  memcpy(buf, passphrase.data(), passphrase.size());
+  return passphrase.size();
+}
+
+class MockHandshakerCallbacks : public Ssl::HandshakerCallbacks {
+public:
+  ~MockHandshakerCallbacks() override{};
+  MOCK_METHOD(bssl::UniquePtr<SSL>, HandOff, (), (override));
+  MOCK_METHOD(void, HandBack, (bssl::UniquePtr<SSL>), (override));
+  MOCK_METHOD(void, OnSuccessCb, (SSL*), (override));
+  MOCK_METHOD(void, OnFailureCb, (), (override));
+};
+
+class HandshakerTest : public SslCertsTest {
+protected:
+  HandshakerTest()
+      : dispatcher_(api_->allocateDispatcher("test_thread")), stream_info_(api_->timeSource()),
+        client_ctx_(SSL_CTX_new(TLS_method())), server_ctx_(SSL_CTX_new(TLS_method())) {
+    // Set up key and cert, initialize two SSL objects and a pair of BIOs for
+    // handshaking.
+    auto key = MakeKey();
+    auto cert = MakeCert();
+    auto chain = std::vector<CRYPTO_BUFFER*>{cert.get()};
+
+    SSL_CTX_set_max_proto_version(server_ctx_.get(), TLS1_2_VERSION);
```

Comment:
Why 1.2? Please add comment.

Reply:
Removed, this turned out not to be necessary.